### PR TITLE
Fix #748: qianfan model throw exception the length of messages must b…

### LINKED ISSUE #748: [BUG] qianfan model throw exception the length of messages must be an odd number
Please provide as much details as possible, this will help us to deliver a fix as soon as possible.
Thank you!

**Describe the bug**
When the total number of Messages in the sliding context window is odd (such as 13), and the messages contain a SystemMessage, the qianfan model will throw an exception, indicating that “the length of messages must be an odd number.”

**Log and Stack trace**
2024-03-14 10:31:18.401 DEBUG 44871 --- [           main] d.l.m.q.c.RequestLoggingInterceptor      : Request:
- method: POST
- url: https://aip.baidubce.com/rpc/2.0/ai_custom/v1/wenxinworkshop/chat/completions_pro?access_token=24.b8fc5f59228672bffa6a9c8af2281d59.2592000.1712975478.282335-56206182
- headers: [Authorization: Bearer CFzzSi9v0Ek5m8SYKYXjZWY1]
- body: {"messages":[{"role":"assistant","content":"很抱歉，根据提供的预订号码123-456，我没有找到任何相关的预订信息。请再次确认您的预订号码是否正确，或者提供您的全名以便我进一步查找。如果您需要进一步操作或有其他问题，请随时告诉我，我将竭诚为您服务。"},{"role":"user","content":"姓名：Klaus，姓氏：Heisler"},{"role":"assistant","content":null,"function_call":{"name":"getBookingDetails","arguments":"{\"bookingNumber\":\"123-456\",\"customerName\":\"Klaus\",\"customerSurname\":\"Heisler\"}"}},{"role":"function","content":"{\n  \"bookingNumber\": \"123-456\",\n  \"bookingFrom\": \"2024-03-15\",\n  \"bookingTo\": \"2024-03-17\",\n  \"customer\": {\n    \"name\": \"Klaus\",\n    \"surname\": \"Heisler\"\n  }\n}","name":"getBookingDetails"},{"role":"assistant","content":"根据您的提供的信息，我找到了预订号码为123-456的预订信息。该预订是由Klaus Heisler先生从2024年3月15日至2024年3月17日进行的。如果您需要进一步操作或有其他问题，请随时告诉我，我将竭诚为您服务。"},{"role":"user","content":"查询我的订单"},{"role":"assistant","content":null,"function_call":{"name":"getBookingDetails","arguments":"{\"bookingNumber\":\"123-456\",\"customerName\":\"Klaus\",\"customerSurname\":\"Heisler\"}"}},{"role":"function","content":"{\n  \"bookingNumber\": \"123-456\",\n  \"bookingFrom\": \"2024-03-15\",\n  \"bookingTo\": \"2024-03-17\",\n  \"customer\": {\n    \"name\": \"Klaus\",\n    \"surname\": \"Heisler\"\n  }\n}","name":"getBookingDetails"},{"role":"assistant","content":"根据您的请求，我已经查询到了您的订单信息。您的预订号码是123-456，预订时间是从2024年3月15日至2024年3月17日，预订人是Klaus Heisler先生。如果您需要进一步操作或有其他问题，请随时告诉我，我将竭诚为您服务。"},{"role":"user","content":"取消订单"},{"role":"assistant","content":null,"function_call":{"name":"cancelBooking","arguments":"{\"bookingNumber\":\"123-456\",\"customerName\":\"Klaus\",\"customerSurname\":\"Heisler\"}"}},{"role":"function","content":"Booking 123-456 cannot be canceled","name":"cancelBooking"}],"temperature":0.7,"functions":[{"name":"getBookingDetails","description":"query detail info of book","parameters":{"type":"object","properties":{"customerSurname":{"type":"string"},"customerName":{"type":"string"},"bookingNumber":{"type":"string"}},"required":["bookingNumber","customerName","customerSurname"]}},{"name":"cancelBooking","description":"cancel booking","parameters":{"type":"object","properties":{"customerSurname":{"type":"string"},"customerName":{"type":"string"},"bookingNumber":{"type":"string"}},"required":["bookingNumber","customerName","customerSurname"]}}],"system":"You are a customer support agent of a car rental company named \u0027Miles of Smiles\u0027.\nBefore providing information about booking or cancelling booking, you MUST always check:\nbooking number, customer name and surname.\nToday is 2024-03-14."}
2024-03-14 10:31:18.490 DEBUG 44871 --- [           main] d.l.m.q.c.ResponseLoggingInterceptor     : Response:
- status code: 200
- headers: [Access-Control-Allow-Headers: Content-Type], [Access-Control-Allow-Origin: *], [Appid: 56206182], [Connection: keep-alive], [Content-Length: 101], [Content-Type: application/json; charset=utf-8], [Date: Thu, 14 Mar 2024 02:31:18 GMT], [Server: Apache], [Set-Cookie: BAIDUID=8C6D33EF609C3771C1A18D4AF7CF229E:FG=1; Path=/; Domain=baidu.com; Max-Age=31536000], [Set-Cookie: BAIDUID_BFESS=8C6D33EF609C3771C1A18D4AF7CF229E:FG=1; Path=/; Domain=baidu.com; Max-Age=31536000; Secure; SameSite=None], [X-Aipe-Self-Def: eb_total_tokens:0,prompt_tokens:0,id:as-cf0t3kc4ch], [X-Baidu-Request-Id: e6122d98b544473a587aef5c089d8bcc1000767], [X-Openapi-Server-Timestamp: 1710383478]
- body: {"error_code":336006,"error_msg":"the length of messages must be an odd number","id":"as-cf0t3kc4ch"}
2024-03-14 10:31:18.496  INFO 44871 --- [           main] ConditionEvaluationReportLoggingListener : 

Error starting ApplicationContext. To display the conditions report re-run your application with 'debug' enabled.
2024-03-14 10:31:18.502 ERROR 44871 --- [           main] o.s.boot.SpringApplication               : Application run failed

java.lang.IllegalStateException: Failed to execute ApplicationRunner
	at org.springframework.boot.SpringApplication.callRunner(SpringApplication.java:773) [spring-boot-2.1.6.RELEASE.jar:2.1.6.RELEASE]
	at org.springframework.boot.SpringApplication.callRunners(SpringApplication.java:760) [spring-boot-2.1.6.RELEASE.jar:2.1.6.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:318) [spring-boot-2.1.6.RELEASE.jar:2.1.6.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1213) [spring-boot-2.1.6.RELEASE.jar:2.1.6.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1202) [spring-boot-2.1.6.RELEASE.jar:2.1.6.RELEASE]
	at cm.xb.langchain4j_test.booking.CustomerSupportApplication.main(CustomerSupportApplication.java:149) [classes/:na]
Caused by: java.lang.IllegalArgumentException: text cannot be null or blank
	at dev.langchain4j.internal.Exceptions.illegalArgument(Exceptions.java:19) ~[langchain4j-core-0.28.0.jar:na]
	at dev.langchain4j.internal.ValidationUtils.ensureNotBlank(ValidationUtils.java:98) ~[langchain4j-core-0.28.0.jar:na]
	at dev.langchain4j.data.message.AiMessage.<init>(AiMessage.java:30) ~[langchain4j-core-0.28.0.jar:na]
	at dev.langchain4j.data.message.AiMessage.from(AiMessage.java:100) ~[langchain4j-core-0.28.0.jar:na]
	at dev.langchain4j.data.message.AiMessage.aiMessage(AiMessage.java:127) ~[langchain4j-core-0.28.0.jar:na]
	at dev.langchain4j.model.qianfan.InternalQianfanHelper.aiMessageFrom(InternalQianfanHelper.java:162) ~[langchain4j-qianfan-0.28.0.jar:na]
	at dev.langchain4j.model.qianfan.QianfanChatModel.generate(QianfanChatModel.java:131) ~[langchain4j-qianfan-0.28.0.jar:na]
	at dev.langchain4j.model.qianfan.QianfanChatModel.generate(QianfanChatModel.java:95) ~[langchain4j-qianfan-0.28.0.jar:na]
	at dev.langchain4j.service.DefaultAiServices$1.invoke(DefaultAiServices.java:165) ~[langchain4j-0.28.0.jar:na]
	at cm.xb.langchain4j_test.booking.$Proxy52.chat(Unknown Source) ~[na:na]
	at cm.xb.langchain4j_test.booking.CustomerSupportApplication.lambda$interactiveChatRunner$0(CustomerSupportApplication.java:78) [classes/:na]
	at org.springframework.boot.SpringApplication.callRunner(SpringApplication.java:770) [spring-boot-2.1.6.RELEASE.jar:2.1.6.RELEASE]
	... 5 common frames omitted

**To Reproduce**
Please provide a relevant code snippets to reproduce this bug.

**Expected behavior**
To ensure that messages are odd in number, remove the oldest message.

**Please complete the following information:**
- LangChain4j version: e.g. 0.28.0
- Java version: e.g. 8
- Spring Boot version (if applicable): e.g. 2.1.6.RELEASE

**Additional context**

### PR DESCRIPTION
**Pull request by yangxb2010000**
Fix #748: qianfan model throw exception the length of messages must be an odd number.

## Summary by CodeRabbit

- **Bug Fixes**
	- Adjusted message processing to ensure message length is always odd, enhancing consistency in data handling.

### CODE REVIEW

File: langchain4j-qianfan/src/main/java/dev/langchain4j/model/qianfan/InternalQianfanHelper.java [hunk 1]
```diff
@@ -179,10 +179,18 @@ static String getSystemMessage(List<ChatMessage> messages) {
...
     }
     public static List<Message> toOpenAiMessages(List<ChatMessage> messages) {
-        return messages.stream()
+         List<Message> aiMessages = messages.stream()
                 .filter(chatMessage -> !(chatMessage instanceof SystemMessage))
                 .map(InternalQianfanHelper::toQianfanMessage)
                 .collect(toList());
+
+         // ensure the length of messages is odd
+        if(aiMessages.size() == 0 || aiMessages.size() % 2 == 1){
```
Comment (langchain4j) [inline langchain4j-qianfan/src/main/java/dev/langchain4j/model/qianfan/InternalQianfanHelper.java:188]: What is the reason for this "must be odd" requirement?

File: langchain4j-qianfan/src/main/java/dev/langchain4j/model/qianfan/InternalQianfanHelper.java [hunk 2]
```diff
@@ -179,10 +179,18 @@ static String getSystemMessage(List<ChatMessage> messages) {
...
+            return aiMessages;
+        }
+
+        aiMessages.remove(0);
```
Comment (langchain4j) [inline langchain4j-qianfan/src/main/java/dev/langchain4j/model/qianfan/InternalQianfanHelper.java:192]: worth adding a log on WARN level? that the message was removed

**Review by ghost — COMMENTED** [review]
@yangxb2010000 thanks!
Please add a test for this case

File: langchain4j-qianfan/src/main/java/dev/langchain4j/model/qianfan/InternalQianfanHelper.java (hunk 1 above)
Comment (yangxb2010000) [reply langchain4j-qianfan/src/main/java/dev/langchain4j/model/qianfan/InternalQianfanHelper.java:188]: If the length of messages is even, the Qianfan model will return an error code 336006. For reference, please see the link:https://cloud.baidu.com/doc/WENXINWORKSHOP/s/tlmyncueh

Comment (langchain4j) [reply langchain4j-qianfan/src/main/java/dev/langchain4j/model/qianfan/InternalQianfanHelper.java:188]: I understand that the error will be returned. But what is behind this requirement? Why can't the length be even?

Comment (yangxb2010000) [reply langchain4j-qianfan/src/main/java/dev/langchain4j/model/qianfan/InternalQianfanHelper.java:188]: I don't know either. I'm just a user, not a developer of the Qianfan model.

**Comment by yangxb2010000** [comment]
> @yangxb2010000 thanks! Please add a test for this case

The testcase has already been added.

File: langchain4j-qianfan/src/main/java/dev/langchain4j/model/qianfan/InternalQianfanHelper.java (hunk 2 above)
Comment (yangxb2010000) [reply langchain4j-qianfan/src/main/java/dev/langchain4j/model/qianfan/InternalQianfanHelper.java:192]: sorry, I couldn't find a log demo in llm model module. May we proceed without it?

**Review by ghost — APPROVED** [review]
@yangxb2010000 thank you!

**Pull request merged**